File: oneflow/core/autograd/gradient_funcs/concat.cpp
/*
Copyright 2020 The OneFlow Authors. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
#include "oneflow/core/framework/op_expr_grad_function.h"
#include "oneflow/core/framework/op_builder.h"
#include "oneflow/core/framework/op_expr.h"
#include "oneflow/core/framework/op_interpreter/op_interpreter_util.h"
#include "oneflow/core/functional/functional.h"

namespace oneflow {
namespace one {

struct ConcatCaptureState : public AutoGradCaptureState {
  std::vector<bool> requires_grad;
  int64_t axis;
  int64_t input_num;
};

class Concat : public OpExprGradFunction<ConcatCaptureState> {
 public:
  Maybe<void> Capture(ConcatCaptureState* state, const TensorTuple& inputs,
                      const TensorTuple& outputs, const OpInterpCtx* ctx) const override;
  Maybe<void> Apply(const ConcatCaptureState* state, const TensorTuple& out_grads,
                    TensorTuple* in_grads) const override;
};

Maybe<void> Concat::Capture(ConcatCaptureState* state, const TensorTuple& inputs,
                            const TensorTuple& outputs, const OpInterpCtx* ctx) const {
  state->requires_grad.resize(inputs.size());
  for (int i = 0; i < inputs.size(); ++i) {
    state->requires_grad[i] = inputs.at(i)->requires_grad();
  }

  auto* interp_ctx = dynamic_cast<const ConcatOpInterpCtx*>(ctx);
  state->axis = interp_ctx->axis;
  for (const auto& input : inputs) { state->SaveTensorForBackward(input); }
  state->input_num = inputs.size();
  return Maybe<void>::Ok();
}

Maybe<void> Concat::Apply(const ConcatCaptureState* state, const TensorTuple& out_grads,
                          TensorTuple* in_grads) const {
  CHECK_EQ_OR_RETURN(out_grads.size(), 1);
  in_grads->resize(state->input_num);
  TensorTuple like(state->input_num);
  for (int i = 0; i < state->input_num; ++i) { like[i] = state->SavedTensors().at(i); }
  const auto& results = JUST(functional::SplitLike(out_grads.at(0), like, state->axis));
  CHECK_EQ_OR_RETURN(results->size(), state->input_num);

  for (int i = 0; i < state->input_num; ++i)
    if (state->requires_grad.at(i)) { in_grads->at(i) = results->at(i); }
  return Maybe<void>::Ok();
}

REGISTER_OP_EXPR_GRAD_FUNCTION("concat", Concat);

}  // namespace one
}  // namespace oneflow

File: python/oneflow/compatible/single_client/nn/modules/concat.py
"""
Copyright 2020 The OneFlow Authors. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""
from typing import Optional, Sequence

from oneflow.compatible import single_client as flow
from oneflow.compatible.single_client.framework.tensor import Tensor, register_tensor_op
from oneflow.compatible.single_client.nn.module import Module


class Cat(Module):
    def __init__(self, dim=0) -> None:
        super().__init__()
        self.axis = dim

    def forward(self, inputs):
        if len(inputs) == 1:
            return inputs[0]
        axis = self.axis
        assert len(inputs) >= 2
        if axis < 0:
            axis += len(inputs[0].shape)
        assert axis >= 0 and axis < len(
            inputs[0].shape
        ), "axis must be in range [0, num_axes of inputs)"
        first_input_shape = inputs[0].shape
        dynamic_dim_size = 0
        for input in inputs:
            assert len(input.shape) == len(first_input_shape)
            for i in range(len(input.shape)):
                if i == axis:
                    dynamic_dim_size += input.shape[i]
                else:
                    assert input.shape[i] == first_input_shape[i]
        return flow.F.concat(inputs, axis=axis, max_dim_size=dynamic_dim_size)


def concat_op(inputs, dim=0):
    """Concatenate two or more `Tensor` s at specified axis.

    Analogous to `numpy.concatenate <https://docs.scipy.org/doc/numpy/reference/generated/numpy.concatenate.html>`_

    Args:
        inputs: a `list` of `Tensor`
        dim: a `int`.

    Returns:
        A `Tensor`

    For example:

    .. code-block:: python

        >>> import oneflow.compatible.single_client.experimental as flow
        >>> flow.enable_eager_execution()
        >>> import numpy as np

        >>> input1 = flow.Tensor(np.random.randn(2, 6, 5, 3), dtype=flow.float32)
        >>> input2 = flow.Tensor(np.random.randn(2, 6, 5, 3), dtype=flow.float32)
        >>> input3 = flow.Tensor(np.random.randn(2, 6, 5, 3), dtype=flow.float32)

        >>> out = flow.cat([input1, input2, input3], dim=1)
        >>> out.shape
        flow.Size([2, 18, 5, 3])

    """
    return Cat(dim=dim)(inputs)


if __name__ == "__main__":
    import doctest

    doctest.testmod(raise_on_error=True)
